kong/plugins/zipkin/opentracing.lua

```diff
@@ -48,23 +46,26 @@ end
 
 function OpenTracingHandler:initialise_request(conf, ctx)
 	local tracer = self:get_tracer(conf)
-	local var = ngx.var
-	local wire_context = tracer:extract("http_headers", ngx.req.get_headers()) -- could be nil
-	local method, request_url -- if reading request line fails then var.request_uri is nil
-	if var.request_uri then
-		method = ngx.req.get_method()
-		request_url = var.scheme .. "://" .. var.host .. ":" .. var.server_port .. var.request_uri
-	end
+	local req = kong.request
+	local wire_context = tracer:extract("http_headers", req.get_headers()) -- could be nil
+	local method, url
+	local path_with_query = req.get_path_with_query()
+	if path_with_query ~= "" then
```

Comment:
From what I understand there are two options here: either the request is a valid http, or it isn't.

* If the request is a valid http request, then `ngx.req.request_uri` always returns at least a slash in the path. We have a test for that particular condition here: https://github.com/Kong/kong/blob/next/t/01-pdk/04-request/09-get_path.t#L32 (`kong.request.get_path` is implemented around `ngx.req.request_uri`)

* If the request is not a valid http request, then `ngx.req.request_uri` will be `nil`, and `kong.request.get_path_with_query` will return `""` because of the `or`

Unless I missed anything regarding how to reproduce this, I think we're safe.

Reply:
> If the request is a valid http request, then ngx.req.request_uri always returns at least a slash in the path.

What about CORS when just `*` is the path?

> Unless I missed anything regarding how to reproduce this, I think we're safe.

I think in HTTP2 you can end up with the path as the empty string.
Infact in HTTP1 depending on how you parse it you might end up with the empty string (in particular it's probably interesting for WAFs)

Reply:
I investigated this more just to be sure. In our tests the path is always at least "/" because test_nginx makes it a default. See for example here:

https://github.com/openresty/test-nginx/blob/master/lib/Test/Nginx/Socket.pm#L126

So my argument that "our tests show it's always non-empty" is not valid - you can't "not specify a request path" using test-nginx.

I also tried looking at the nginx source code but honestly I could not make sense about how the `$request_uri` env var is initialized (too many [layers of abstraction on top of C](https://github.com/nginx/nginx/blob/4bf4650f2f10f7bbacfe7a33da744f18951d416d/src/http/ngx_http_variables.c#L218) for me to make sense of it quickly).

I think I found an easier answer somewhere else: in Kong's nginx conf. Our own proxy declaration has a `location /` right there when declaring the proxy:

https://github.com/Kong/kong/blob/master/kong/templates/nginx_kong.lua#L115

If we're using `location /` when declaring a proxy, any request accepted by kong should have at least a `/` path.

Which brings me to my next point: I don't understand how a *non*-valid http request can possibly "pass through" that. The way the nginx config file is structured, non-http requests should be rejected at the nginx level, before they even reach kong. Unless there is some kind of custom nginx templating involved, or I missed something else.

Reply:
> What about CORS when just * is the path?

The main reason behind this PR is making the plugin compatible with the version of kong that we have on `next`. I think that one only supports http with a path (see my previous comment about the `location /` section).

I am not opposed to revisiting this in the future, but it seems that covering those cases falls out of the scope of this particular PR.

When that time comes, I would hope that kong has something more explicit than checking `kong.request_uri` for `nil`. Something like:

```
if kong.request.is_cors() then
  do CORS stuff
elseif kong.request.is_http() then
  do http stuff
else ...
end
```

> I think in HTTP2 you can end up with the path as the empty string.
Infact in HTTP1 [...] interesting for WAFs)

Same answer: I just need this to support kong's `next` branch now. If someone is using a custom nginx template to support the scenarios you mention, I am ok with letting them deal with those cases manually for now (I'm not opposed to supporting them in the future, when we accept them in Kong. But not for this PR).